File: apps/api/models/stream.py
"""Pydantic models for the class stream (posts)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from urllib.parse import urlparse

from pydantic import BaseModel, field_validator

PostType = Literal["announcement", "material"]


class ClassPost(BaseModel):
    id: str
    class_id: str
    author_id: str
    type: PostType
    title: str | None = None
    body: str | None = None
    link_url: str | None = None
    created_at: datetime | str | None = None


class CreatePostRequest(BaseModel):
    type: PostType
    title: str | None = None
    body: str | None = None
    link_url: str | None = None

    @field_validator("link_url")
    @classmethod
    def _validate_link_url(cls, v: str | None) -> str | None:
        if v is None:
            return v
        parsed = urlparse(v)
        if parsed.scheme not in ("http", "https"):
            raise ValueError("link_url must use http or https scheme")
        return v

File: apps/web/lib/stream.ts
import { apiFetch } from "@/lib/api";
import { shortId } from "@/lib/classroom";

export type PostType = "announcement" | "material";

export interface ClassPost {
  id: string;
  class_id: string;
  author_id: string;
  type: PostType;
  title?: string | null;
  body?: string | null;
  link_url?: string | null;
  created_at?: string | null;
}

export interface CreatePostRequest {
  type: PostType;
  title?: string | null;
  body?: string | null;
  link_url?: string | null;
}

export async function fetchPosts(classId: string): Promise<ClassPost[]> {
  return apiFetch<ClassPost[]>(`/api/classes/${classId}/posts`);
}

export async function createPost(
  classId: string,
  req: CreatePostRequest,
): Promise<ClassPost> {
  return apiFetch<ClassPost>(`/api/classes/${classId}/posts`, {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export async function deletePost(
  classId: string,
  postId: string,
): Promise<void> {
  await apiFetch<void>(`/api/classes/${classId}/posts/${postId}`, {
    method: "DELETE",
  });
}

/** Returns a short fallback display name from a UUID. */
export function authorShortId(userId: string): string {
  return shortId(userId);
}

/** Format a date string as a relative or absolute label. */
export function formatPostDate(value?: string | null): string {
  if (!value) return "";
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return value;
  const diff = Date.now() - date.getTime();
  const minutes = Math.floor(diff / 60_000);
  if (minutes < 1) return "just now";
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ago`;
  return new Intl.DateTimeFormat(undefined, {
    month: "short",
    day: "numeric",
  }).format(date);
}

/** Regex for a valid YouTube video ID (exactly 11 URL-safe chars). */
const YOUTUBE_ID_RE = /^[a-zA-Z0-9_-]{11}$/;

/** Detect a YouTube URL and return the video ID, or null. */
export function youtubeVideoId(url: string): string | null {
  try {
    const parsed = new URL(url);
    let id: string | null = null;
    if (
      parsed.hostname === "www.youtube.com" ||
      parsed.hostname === "youtube.com"
    ) {
      id = parsed.searchParams.get("v");
    } else if (parsed.hostname === "youtu.be") {
      id = parsed.pathname.slice(1) || null;
    }
    if (id && YOUTUBE_ID_RE.test(id)) return id;
  } catch {
    // Not a valid URL.
  }
  return null;
}
